### PR TITLE
 API Widgets and WidgetAreas are versioned and are owned by pages

### LINKED ISSUE #148: ENHANCEMENT Apply Versioned extension to Widget model (v4 only)
With the new ownership model in place, it should be simple to apply versioned and and ownership relationship between pages and widgets.

https://docs.silverstripe.org/en/4/developer_guides/model/versioning/#dataobject-ownership

Replaces https://github.com/silverstripe/silverstripe-cms/issues/322

### PR DESCRIPTION
**Pull request by robbieaverill**
- [x] **Note:** contains the commits from #162 which needs to be merged first.

This pull request changes the following:

* Array syntax
* Remove pointless doc blocks (where inherited ones would be more descriptive)
* Add Versioned to WidgetArea and Widget to track changes
* Implement Page ownership of WidgetArea and WidgetArea ownership of Widgets, with cascading deletes (removed custom code for this)
* Moved WidgetController into Widget's namespace since it's expected to be there to work (see `Widget::getController`)

Resolves #148

### CODE REVIEW

**Comment by tractorcow** [comment]
Can you please add a comment on the justification for versioned without stages vs versioned with stages?

**Comment by robbieaverill** [comment]
It's primarily because the widget area editor user interface doesn't have save/publish buttons in it at the moment, and I didn't want to add them right now

**Comment by robbieaverill** [comment]
It's primarily because the widget area editor user interface doesn't have save/publish buttons in it at the moment, and I didn't want to add them right now

File: src/Model/WidgetArea.php [hunk 1]
```diff
@@ -2,32 +2,36 @@
...
+    private static $owns = [
+        'Widgets',
+    ];
+
+    private static $cascade_deletes = [
+        'Widgets',
+    ];
+
+    private static $extensions = [
+        Versioned::class . '.versioned',
```
Comment (NightJar) [inline src/Model/WidgetArea.php]: Seems like this should just be `Versioned::class,`

File: src/Model/Widget.php [hunk 2]
```diff
@@ -76,18 +63,16 @@ class Widget extends DataObject
...
-    );
+    ];
 
-    /**
-     * @var string
-     */
     private static $table_name = 'Widget';
 
+    private static $extensions = [
+        Versioned::class . '.versioned',
```
Comment (NightJar) [inline src/Model/Widget.php]: Remove the ` . '.versioned'` as discussed with @tractorcow - this is the wrong kind of versioning for this application

File: src/Model/WidgetController.php [hunk 3]
```diff
@@ -1,12 +1,11 @@
 <?php
 
-namespace SilverStripe\Widgets\Controllers;
+namespace SilverStripe\Widgets\Model;
```
Comment (NightJar) [inline src/Model/WidgetController.php]: Why was this move done?
Naming convention sees it as an oxymoron :/

File: src/Model/Widget.php [hunk 4]
```diff
@@ -267,10 +252,13 @@ public function getController()
         }
 
         if (!class_exists($controllerClass)) {
-            throw new Exception('Could not find controller class for ' . $controllerClass);
+            throw new Exception('Could not find controller class for ' . static::class);
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Model/Widget.php]: The idea is to say "we couldn't find a controller for MyCustomWidget"

File: src/Model/WidgetController.php (hunk 3 above)
Comment (robbieaverill) [reply src/Model/WidgetController.php]: The controller must be in the same namespace, see `Widget::getController` for the current lookup logic

**Comment by robbieaverill** [comment]
Ok, I've updated the versioning implementation.

The reason for this change is so that you can have draft versions of a widget without them being shown immediately on the live site.

Example:

* Create a page, add a widget
* Set some content in the widget
* Publish the page
* Change the widget contents
* Save the page as draft (don't publish)

Live site will respect the published state of the widget rather than show the new draft content as it would have previously.

**Comment by tractorcow** [comment]
Thanks @robbieaverill great job. :)

File: src/Model/Widget.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (NightJar) [reply src/Model/Widget.php]: Right, maybe it's me getting confused between `self::class`, `static::class`, `get_class()`, and `__CLASS__`, but I feel like the difference here is that `static::class` will be _this_ class, i.e. `Widget` or a subclass there of.

Where as the variable here being used in the `if` condition will be for a corresponding controller, a subclass of `WidgetController`.

tl;dr, I think this error message will give the wrong information.

[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Model/Widget.php]: Perhaps it needs a better doc block.

What the method is doing is trying to find an appropriate controller for the current Widget instance. It does this by looking backwards through the inheritance chain until it finds a controller class that exists.

This error is thrown at the end when it couldn't find any controllers at all, and the `$controllerClass` at this point will be `SilverStripe\View\ViewableDataController` which is misleading and incorrect.

I changed it to include the late static bound (`static::class`) class name of the widget that threw the error, so that when you use the CMS the error you get will say "we couldn't find a controller for YourCustomWidgetName", rather than some low level class that has nothing to do with the widgets in this context.

Does this make sense?

[the remark this answers is not in the document]
Comment (NightJar) [reply src/Model/Widget.php]: Yep, perfect sense :)
Thanks for explaining that 👍
